language/move-lang/src/parser/merge_spec_modules.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Merges specification modules into their target modules.
+//!
+//! There are some issues with this approach which we may want to fix down the road:
+//! - If a spec module contains a `use`, we don't want the target module be able to use it.
+//! - Similarly, we also *may* want the spec module not be able to see target module `use`
+//!   declarations, and require it to repeat them.
+//! A solution to both problems can be to mark names introduced by `use` to whether they
+//! are for specs or not, and allow the older to resolve only in spec contexts.
+
+use crate::{
+    parser::ast::{Definition, ModuleDefinition, Program},
+    shared::*,
+};
+use std::collections::BTreeMap;
+
+/// Given a parsed program, merge all specification modules into their target modules.
+pub fn program(compilation_env: &mut CompilationEnv, prog: Program) -> Program {
+    let Program {
+        source_definitions,
+        lib_definitions,
+    } = prog;
+
+    // Phase 1: extract all spec modules.
+    let mut spec_modules = BTreeMap::new();
+    let mut source_definitions = extract_spec_modules(source_definitions, &mut spec_modules);
+    let mut lib_definitions = extract_spec_modules(lib_definitions, &mut spec_modules);
+
+    // Phase 2: Go over remaining proper modules and merge spec modules.
+    merge_spec_modules(&mut source_definitions, &mut spec_modules);
+    merge_spec_modules(&mut lib_definitions, &mut spec_modules);
+
+    // Remaining spec modules could not be merged, report errors.
+    for (_, m) in spec_modules {
+        compilation_env.add_error(vec![(
+            m.name.loc(),
+            "Cannot associate specification with any target module in this compilation. \
+                    A module specification cannot be compiled standalone."
+                .to_owned(),
+        )]);
+    }
+    Program {
+        source_definitions,
+        lib_definitions,
+    }
+}
+
+fn extract_spec_modules(
+    defs: Vec<Definition>,
+    spec_modules: &mut BTreeMap<(Address, String), ModuleDefinition>,
+) -> Vec<Definition> {
+    use Definition::*;
+    defs.into_iter()
+        .filter_map(|def| match def {
+            Module(m) => extract_spec_module(&None, m, spec_modules).map(Module),
+            Address(attrs, loc, addr, module_defs) => {
+                let addr_opt = Some(addr);
+                let module_defs = module_defs
+                    .into_iter()
+                    .filter_map(|m| extract_spec_module(&addr_opt, m, spec_modules))
+                    .collect::<Vec<_>>();
+                if module_defs.is_empty() {
+                    None
```

Comment:
It is fine to have an empty address decl, especially once named address syntax is in, so the address should also be kept

Reply:
Done